### PR TITLE
coprocessor/statistics: push down AnalyzeReq

### PR DESCRIPTION
**Pull request by AndreMouche**
@BusyJay @hicqu @lamxTyler PTAL

### CODE REVIEW

File: src/coprocessor/codec/table.rs [hunk 1]
```diff
@@ -298,6 +298,21 @@ impl RowColsDict {
         self.value.append(value);
         self.cols.insert(cid, RowColMeta::new(offset, length));
     }
+
+    pub fn get_binary(&self) -> &[u8] {
```
Comment (BusyJay) [inline src/coprocessor/codec/table.rs]: Please add comment.

File: src/coprocessor/statistics/analyze.rs [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+use coprocessor::dag::executor::{Executor, IndexScanExecutor, TableScanExecutor};
+use coprocessor::endpoint::ReqContext;
+use coprocessor::codec::datum;
+use coprocessor::{Error, Result};
+use storage::{Snapshot, SnapshotStore, Statistics};
+use super::fmsketch::FMSketch;
+use super::histogram::Histogram;
+
+pub struct AnalyzeContext<'a> {
```
Comment (BusyJay) [inline src/coprocessor/statistics/analyze.rs]: Please add comment.

**Comment by shenli** [comment]
@BusyJay PTAL

File: src/coprocessor/statistics/analyze.rs [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        let mut scanner = IndexScanExecutor::new_with_cols_len(
+            req.get_num_columns() as i64,
+            self.ranges,
+            self.snap,
+            self.statistics,
+        );
+        let mut hist = Histogram::new(req.get_bucket_size() as usize);
+        while let Some(row) = try!(scanner.next()) {
+            let bytes = row.data.get_binary().to_vec();
+            hist.append(bytes);
```
Comment (BusyJay) [inline src/coprocessor/statistics/analyze.rs]: Can it accept a `&[u8]` instead?

File: src/coprocessor/statistics/analyze.rs [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+    }
+
+    // `collect_samples_and_estimate_ndvs` returns the sample collectors which contain total count,
+    // null count and distinct values count. And it also returns the statistic builder for PK
+    // which contains the histogram. See https://en.wikipedia.org/wiki/Reservoir_sampling
+    fn collect_samples_and_estimate_ndvs(mut self) -> Result<(Vec<SampleCollector>, Histogram)> {
+        let mut pk_builder = if self.pk_id == -1 {
+            Histogram::default()
+        } else {
+            Histogram::new(self.max_bucket_size)
```
Comment (BusyJay) [inline src/coprocessor/statistics/analyze.rs]: Why?

File: src/coprocessor/statistics/analyze.rs [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+            self.null_count += 1;
+            return;
+        }
+        self.count += 1;
+        self.sketch.insert(&data);
+        if self.samples.len() < self.max_sample_size {
+            self.samples.push(data);
+            return;
+        }
+        let seed: u64 = rand::random();
```
Comment (BusyJay) [inline src/coprocessor/statistics/analyze.rs]: Cache the random object instead.

File: src/coprocessor/statistics/analyze.rs (hunk 4 above)
Comment (hicqu) [reply src/coprocessor/statistics/analyze.rs]: I will figure out it later.

File: src/coprocessor/statistics/analyze.rs [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+            .collect();
+        while let Some(row) = try!(self.data.next()) {
+            let cols = try!(row.get_binary_cols(&self.cols));
+            let mut cols_iter = cols.into_iter();
+            if self.pk_id != -1 {
+                if let Some(v) = cols_iter.next() {
+                    pk_builder.append(&v);
+                }
+            }
+            for (id, val) in cols_iter.enumerate() {
```
Comment (BusyJay) [inline src/coprocessor/statistics/analyze.rs]: You can use `cols_iter.zip(&mut collectors)`

File: src/coprocessor/statistics/analyze.rs [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    fn new(
+        mut req: AnalyzeColumnsReq,
+        snap: SnapshotStore<'a>,
+        ranges: Vec<KeyRange>,
+        statistics: &'a mut Statistics,
+    ) -> SampleBuilder<'a> {
+        let cols_info = req.take_columns_info();
+        let mut pk_id = -1;
+        let mut num_cols = cols_info.len();
+        if cols_info[0].get_pk_handle() {
```
Comment (BusyJay) [inline src/coprocessor/statistics/analyze.rs]: Can `num_cols == 0`?

File: src/coprocessor/statistics/analyze.rs [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+}
+
+struct SampleBuilder<'a> {
+    data: TableScanExecutor<'a>,
+    cols: Vec<ColumnInfo>,
+    // the number of columns need to be sampled
+    col_len: usize,
+    // If primary key is handle, the pk_id is the id of the primary key.
+    // If not exists, it is -1.
+    pk_id: i64,
```
Comment (BusyJay) [inline src/coprocessor/statistics/analyze.rs]: Is `pk_id` necessary? Can we use `self.col_len != self.cols.len()` instead?

File: src/coprocessor/statistics/analyze.rs [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        }
+        self.count += 1;
+        self.sketch.insert(&data);
+        if self.samples.len() < self.max_sample_size {
+            self.samples.push(data);
+            return;
+        }
+        let seed = self.rng.next_u64();
+        if seed % self.count < self.max_sample_size as u64 {
+            let idx = self.rng.next_u64() as usize % self.max_sample_size;
```
Comment (BusyJay) [inline src/coprocessor/statistics/analyze.rs]: You can use `self.rng.gen_range(0, self.max_sample_size)` here.

File: src/coprocessor/codec/table.rs [hunk 10]
```diff
@@ -298,6 +298,19 @@ impl RowColsDict {
...
+    pub fn get_column_values(&self) -> &[u8] {
+        let mut start = self.value.len();
+        let mut length = 0;
+        for meta in self.cols.values() {
+            if meta.offset < start {
+                start = meta.offset;
+            }
+            length += meta.length;
+        }
+        &self.value[start..(start + length)]
```
Comment (BusyJay) [inline src/coprocessor/codec/table.rs]: `()` is unnecessary.

File: tests/coprocessor/test_analyze.rs [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+    let req = new_analyze_column_req(&product.table, 3, 3, 3);
+    let resp = handle_request(&end_point, req);
+    assert!(resp.get_data().is_empty(), "{:?}", resp);
+    assert!(resp.has_locked(), "{:?}", resp);
+
+    // with isolation level rc
+    let mut req = new_analyze_column_req(&product.table, 3, 3, 3);
+    let mut ctx = Context::new();
+    ctx.set_isolation_level(IsolationLevel::RC);
+    req.set_context(ctx);
```
Comment (BusyJay) [inline tests/coprocessor/test_analyze.rs]: What's the point setting this?

File: src/coprocessor/statistics/analyze.rs (hunk 4 above)
Comment (hicqu) [reply src/coprocessor/statistics/analyze.rs]: Here `Histogram::new(max_size)` is more normal. `self.pk_id == -1` is a special case which means **PK is not handle** and in this case `Histogram` is just a place holder for build response.

**Comment by hicqu** [comment]
@BusyJay , PTAL, thanks.

File: tests/coprocessor/test_analyze.rs [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    let req = new_analyze_index_req(&product.table, 3, product.name.index);
+    let resp = handle_request(&end_point, req);
+    assert!(resp.get_data().is_empty(), "{:?}", resp);
+    assert!(resp.has_locked(), "{:?}", resp);
+
+    // with isolation level rc
+    let mut req = new_analyze_index_req(&product.table, 3, product.name.index);
+    let mut ctx = Context::new();
+    ctx.set_isolation_level(IsolationLevel::RC);
+    req.set_context(ctx);
```
Comment (BusyJay) [inline tests/coprocessor/test_analyze.rs]: This is useless too.

File: src/coprocessor/statistics/analyze.rs [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+
+use coprocessor::dag::executor::{Executor, IndexScanExecutor, TableScanExecutor};
+use coprocessor::endpoint::ReqContext;
+use coprocessor::codec::datum;
+use coprocessor::{Error, Result};
+use storage::{Snapshot, SnapshotStore, Statistics};
+use super::fmsketch::FMSketch;
+use super::histogram::Histogram;
+
+// `AnalyzeContext` is used to handle `anlyzeReq`
```
Comment (lamxTyler) [inline src/coprocessor/statistics/analyze.rs]: `anlyzeReq` -> `analyzeReq`

**Comment by hicqu** [comment]
@lamxTyler , PTAL again, thanks.

**Comment by lamxTyler** [comment]
LGTM

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**